statsmodels/tools/tools.py

```diff
@@ -233,107 +233,70 @@ def categorical(data, col=None, dictnames=False, drop=False, ):
         data without adding another constant. If 'raise', will raise an
         error if a constant is present. Using 'add' will duplicate the
-        constant, if one is present. Has no effect for structured or
-        recarrays. There is no checking for a constant in this case.
+        constant, if one is present.
 
     Returns
     -------
     data : array
-        The original array with a constant (column of ones) as the first or
+        The original values with a constant (column of ones) as the first or
         last column.
-    '''
-    if _is_using_pandas(data, None):
-        # work on a copy
-        return _pandas_add_constant(data.copy(), prepend, has_constant)
+
+    """
+    is_recarray = _is_recarray(data)
+    if _is_using_pandas(data, None) or is_recarray:
+        from pandas import Series, DataFrame
+
+        if is_recarray:
+            data = DataFrame.from_records(data)
+        elif isinstance(data, Series):
+            data = DataFrame(data)
+        else:
+            data = data.copy()
     else:
         data = np.asarray(data)
-    if not data.dtype.names:
-        var0 = data.var(0) == 0
-        if np.any(var0):
-            if has_constant == 'raise':
-                raise ValueError("data already contains a constant.")
-            elif has_constant == 'skip':
-                return data
-            elif has_constant == 'add':
-                pass
-            else:
-                raise ValueError("Option {0} not understood for "
-                                 "has_constant.".format(has_constant))
-        data = np.column_stack((data, np.ones((data.shape[0], 1))))
+
+    var0 = data.var(0) == 0
+    if np.any(var0):
+        if has_constant == 'raise':
+            raise ValueError("data already contains a constant.")
+        elif has_constant == 'skip':
+            return data
+        elif has_constant == 'add':
+            pass
+        else:
+            raise ValueError("Option {0} not understood for "
+                             "has_constant.".format(has_constant))
+
+    const = np.ones((data.shape[0], 1))
+    if _is_using_pandas(data, None):
         if prepend:
-            return np.roll(data, 1, 1)
+            data.insert(0, 'const', const)
+        else:
+            data['const'] = const
     else:
-        return_rec = data.__class__ is np.recarray
+        data = np.column_stack((data, np.ones((data.shape[0], 1))))
         if prepend:
-            ones = np.ones((data.shape[0], 1), dtype=[('const', float)])
-            data = nprf.append_fields(ones, data.dtype.names,
-                                      [data[i] for i in data.dtype.names],
-                                      usemask=False, asrecarray=return_rec)
-        else:
-            data = nprf.append_fields(data, 'const', np.ones(data.shape[0]),
-                                      usemask=False, asrecarray=return_rec)
+            return np.roll(data, 1, 1)
```

Comment:
I think you should use the explicit `if prepend` paths.
`roll` will need to make a copy, or linalg will need to make a copy when it requires contiguous memory
also `column_stack` creates Fortran ordered arrays which should be better for the Fortran linalg calls, although we never verify this.